File: 312_Burst_Balloons.py
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        nums = [1] + nums + [1]
        size = len(nums)
        dp = [[0] * size for _ in range(size)]
        for k in range(2, size):
            for left in range(0, size - k):
                right = left + k
                for i in range(left + 1, right):
                    dp[left][right] = max(dp[left][right], 
                            nums[left] * nums[i] * nums[right] + dp[left][i] + dp[i][right])
        return dp[0][-1]

File: 312_Burst_Balloons.cpp
class Solution {
public:
    int maxCoins(vector<int>& nums) {
        nums.insert(nums.begin(), 1);
        nums.push_back(1);
        int size = nums.size();
        vector<vector<int> > dp (size, vector<int> (size, 0));
        for (int k = 2; k < size; ++k) {
            for (int left = 0; left < size - k; ++left) {
                int right = left + k;
                for (int i = left + 1; i < right; ++i) {
                    dp[left][right] = max(dp[left][right],
                                          nums[left] * nums[i] * nums[right] 
                                          + dp[left][i] + dp[i][right]);
                }
            }
        }
        return dp[0][size - 1];
    }
};
